ai_oca_bridge_extra_parameters/models/ai_extra_parameter.py

```diff
@@ -0,0 +1,510 @@
+import ast
+import re
+
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+from odoo.tools.safe_eval import safe_eval
+
+
+class AIExtraParameter(models.Model):
+    _name = "ai.extra.parameter"
+    _description = "AI Bridge Extra Parameter"
+
+    name = fields.Char(required=True)
+    expression = fields.Char()
+    formula = fields.Text(
+        default="if object:\n" "    result = {'result_key': object.property}\n",
+    )
+    parameter_type = fields.Selection(
+        [
+            ("record", "Record"),
+            ("self", "Bridge Self"),
+        ],
+        required=True,
+        default="record",
+        help="Defines the context in which this parameter is evaluated. "
+        "'record' means it will be evaluated in the context of the record being processed, "
+        "'self' means it will be evaluated in the context of the AI bridge itself.",
+    )
+
+    evaluate_type = fields.Selection(
+        [
+            ("expression", "Expression"),
+            ("formula", "Formula"),
+        ],
+        required=True,
+        default="expression",
+        help="Defines how the parameter is evaluated. "
+        "'expression' means it will be evaluated as a Python expression, "
+        "'formula' means it will be evaluated as a formula expression.",
+    )
+    FORBIDDEN_PATTERNS = [
+        r"\.cr\b",
+        r"\._cr\b",
+        r"\.execute\s*\(",
+        r"\.commit\s*\(",
+        r"\.rollback\s*\(",
+        r"\.sudo\s*\(",
+        r"\.with_context\s*\(",
+        r"\.write\s*\(",
+        r"\.unlink\s*\(",
+        r"\.invalidate_cache\s*\(",
+        r"\.flush\s*\(",
+        r"\.flush_all\s*\(",
+        r"\bimport\b",
+        r"\b(exec|eval)\s*\(",
+        r"\bopen\s*\(",
+        r"\bgetattr\s*\(",
+        r"__\w+__",
+    ]
+
+    @staticmethod
+    def _extract_variables(expr):
```

Comment:
the methods: "_extract_variables" & "_extract_variables_from_code" are more similars, can u join them?